src/ring_theory/polynomial/cyclotomic.lean

```diff
@@ -0,0 +1,452 @@
+  { use X - 1,
+    simp only [hone, cyclotomic'_one K, set.mem_univ, pnat.one_coe, subsemiring.coe_top,
+    eq_self_iff_true, map_X, map_one, ring_hom.coe_of, and_self, map_sub] },
+  let B : polynomial K := ∏ i in nat.proper_divisors k, cyclotomic' i K,
+  have Bmo : B.monic,
+  { apply monic_prod_of_monic,
+    intros i hi,
+    exact (cyclotomic'.monic i K) },
+  have Bint : B ∈ lifts (int.cast_ring_hom K),
+  { refine subsemiring.prod_mem (lifts (int.cast_ring_hom K)) _,
+    intros x hx,
+    have xsmall := (nat.mem_proper_divisors.1 hx).2,
+    obtain ⟨d, hd⟩ := (nat.mem_proper_divisors.1 hx).1,
+    rw [mul_comm] at hd,
+  exact hk x xsmall (is_primitive_root.pow hpos hzeta hd) },
+  replace Bint := lifts_and_degree_eq_and_monic Bint Bmo,
+  obtain ⟨B₁, hB₁, hB₁deg, hB₁mo⟩ := Bint,
+  let Q₁ : polynomial ℤ := (X ^ k - 1) /ₘ B₁,
+  have huniq : 0 + B * cyclotomic' k K = X ^ k - 1 ∧ (0 : polynomial K).degree < B.degree,
+  { split,
+    { rw [zero_add, mul_comm, ←(prod_cyclotomic'_eq_X_pow_sub_one hpos hzeta),
+      nat.divisors_eq_proper_divisors_insert_self_of_pos hpos],
+      simp only [true_and, finset.prod_insert, not_lt, nat.mem_proper_divisors, dvd_refl] },
+    rw [degree_zero, bot_lt_iff_ne_bot],
+    intro habs,
+    exact (monic.ne_zero Bmo) (degree_eq_bot.1 habs) },
+  replace huniq := div_mod_by_monic_unique (cyclotomic' k K) (0 : polynomial K) Bmo huniq,
+  simp only [lifts, ring_hom.coe_of, ring_hom.mem_srange],
+  use Q₁,
+  rw [(map_div_by_monic (int.cast_ring_hom K) hB₁mo), hB₁, ← huniq.1],
+  simp
+end
+
+/-- If `K` is of characteristic `0` and there is a primitive `n`-th root of unity in `K`,
+then `cyclotomic n K` comes from a unique polynomial with integer coefficients. -/
+lemma unique_int_coeff_of_cycl [char_zero K] {ζ : K} {n : ℕ+} (h : is_primitive_root ζ n) :
+  (∃! (P : polynomial ℤ), map (int.cast_ring_hom K) P = cyclotomic' n K) :=
+begin
+  obtain ⟨P, hP⟩ := int_coeff_of_cyclotomic h,
+  rw exists_unique,
+  use [P, hP.1],
+  intros Q hQ,
+  have mapinj : function.injective (map (int.cast_ring_hom K)),
+  { apply map_injective,
+    simp only [int.cast_injective, int.coe_cast_ring_hom] },
+  rw [function.injective] at mapinj,
+  apply mapinj,
+  rw [hP.1, hQ]
+end
+
+end field
+
+end cyclotomic'
+
+section cyclotomic
+
+/-- The `n`-th cyclotomic polynomial with coefficients in `R`. -/
+def cyclotomic (n : ℕ) (R : Type*) [ring R] : polynomial R :=
+if h : n = 0 then 1 else map (int.cast_ring_hom R) (classical.some (int_coeff_of_cyclotomic
+  (complex.is_primitive_root_exp n h)))
+
+lemma int_cyclotomic_rw {n : ℕ} (h : n ≠ 0) :
+  cyclotomic n ℤ = (classical.some (int_coeff_of_cyclotomic (complex.is_primitive_root_exp n h))) :=
+begin
+  simp only [cyclotomic, h, dif_neg, not_false_iff],
+  ext i,
+  simp only [coeff_map, int.cast_id, ring_hom.eq_int_cast]
+end
+
+/-- `cyclotomic n R` comes from `cyclotomic n ℤ`. -/
+lemma cyclotomic_from_int_cyclotomic (n : ℕ) (R : Type*) [ring R] :
```

Comment:
```suggestion
lemma map_cyclotomic_int (n : ℕ) (R : Type*) [ring R] :
```
Maybe swap the two sides. Also, I think I would put this next to `map_cyclotomic`.